controller/internal/enforcer/applicationproxy/http/http.go

```diff
@@ -155,6 +156,29 @@ func (p *Config) newBaseTLSConfig() *tls.Config {
 	}
 }
 
+// newBaseTLSClientConfig creates the new basic TLS configuration for the client.
+func (p *Config) newBaseTLSClientConfig() *tls.Config {
+	return &tls.Config{
+		GetCertificate:           p.GetCertificateFunc(),
+		NextProtos:               []string{"h2"},
+		PreferServerCipherSuites: true,
+		SessionTicketsDisabled:   true,
+		GetClientCertificate:     p.GetClientCertificateFunc(),
+		// for now lets make it TLS1.2 as supported max Version.
+		// TODO: Need to test before enabling TLS 1.3, currently TLS 1.3 doesn't work with envoy.
+		MaxVersion: tls.VersionTLS12,
+	}
+}
+
+func (p *Config) GetClientCertificateFunc() func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
+	return func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
```

Comment:
if I see this right, there is no need to return a closure here. Just adjust your function with the right function signature and pass it directly in 166 ... same should be done for `GetCertificateFunc` I guess

Reply:
also ... as `cert` is  a field that gets updated, you need  a read lock here as well if I'm not mistaken

Reply:
for #2 I agree, I will make that change.

For #1 The function is going to get bigger once Envoy changes comes in, so I thought of having it in a different function.

Reply:
for pt. 1, this is what I'm saying: change the function signature to:
```
func (p *Config) GetClientCertificateFunc() (*tls.Certificate, error)
```
and assign it like this:
```
GetClientCertificate:     p.GetClientCertificateFunc,
```

Then you can remove the wrapper which builds a closure on every call